feat: use TryLock, add supplementary comments

winterjihwan opened the pull request.

winterjihwan committed:
feat: push ext mutex inwards

winterjihwan committed:
feat: utilize TryLock for non blocking lock operation

winterjihwan committed:
feat: remove box, add comments

winterjihwan committed:
chore: remove unnecessary locks

winterjihwan committed:
comment: add supplementary comments

winterjihwan committed:
comment: more comments

djm07073 commented on crates/revmc-worker/src/error/ext.rs:
```diff
@@ -14,9 +14,15 @@ pub enum ExtError {
+    RwLockPoison { err: String },
+
+    #[error("Mutex poison error: {err}")]
+    MutexPoison { err: String },
```
it seems not used

winterjihwan replied:
done

djm07073 commented on crates/revmc-worker/src/worker/sleddb.rs:
```diff
@@ -6,6 +6,7 @@ use std::sync::Arc;
 
 use super::path::sleddb_path;
 
+/// In-Memory Database for local storage of compiled multiple ExternalFns
```
is in-memory database right expression? how about embedded database?

winterjihwan replied:
You are right, changed it to embedded

djm07073 commented on crates/revmc-worker/src/tests/worker.rs:
```diff
@@ -46,15 +41,15 @@ fn fib_call_data() -> Bytes {
 #[test]
 fn test_compiler_cache_retrieval() {
-    let ext_worker = EXTCompileWorker::new(1, 3, 128);
+    let ext_worker = Arc::new(EXTCompileWorker::new(1, 3, 128));
```
can we move wrapping with Arc in EXTCompilerWorker::new? instead of doing outside. we should force to use arc to ensure thread-safe

winterjihwan replied:
Okay, but I'm not sure if it's a good practice

winterjihwan replied:
done

winterjihwan replied:
+ the compiler will complain anyway if the user do not wrap it with Arc when using it for multithreading

djm07073 replied:
yeah, so it is good to force wrapping arc pointer in creating instance

winterjihwan replied:
okay

djm07073 commented on crates/revmc-worker/src/external.rs:
```diff
@@ -16,53 +23,72 @@ pub(crate) static SLED_DB: OnceCell<Arc<RwLock<SledDB<B256>>>> = OnceCell::new()
+                write_lock.put(code_hash, (f, lib));
 
-                if let Some((f, _)) = self.cache.get(&code_hash) {
+                if let Some((f, _)) = write_lock.get(&code_hash) {
```
we don't have to get function from cache. just return f

```suggestion
    return Ok(Some(*f));
```

winterjihwan replied:
the purpose of this is to check if function exists in cache

djm07073 replied:
if there are any failed cases of putting f, lib in cache? if there are not, i think we don't have to check it

djm07073 commented on crates/revmc/src/compiler/mod.rs:
```diff
@@ -605,3 +605,4 @@ mod default_attrs {
         (std::mem::size_of::<T>(), std::mem::align_of::<T>())
     }
 }
+
```
delete unneccesary line

winterjihwan replied:
done

djm07073 commented on crates/revmc-worker/src/handler.rs:
```diff
@@ -1,48 +1,46 @@
-                    .unwrap()
-                    .work(spec_id, code_hash, bytecode.original_bytes());
 
+                if let Err(_err) =
+                    context.external.work(spec_id, code_hash, bytecode.original_bytes())
+                {
+                };
```
we don't have to handling error from exteranl.work?

winterjihwan replied:
it should be done in revm-wise, since register handler is being called in revm
previously just logged it with tracer, but since we don't use tracer, i think it's okay to leave it as a placeholder

djm07073 replied:
okay i'll add error log to trace

djm07073 commented on crates/revmc-worker/src/handler.rs:
```diff
@@ -1,48 +1,46 @@
-                if let Err(err) = &res {
-                    tracing::error!("Extern Fn Call: with bytecode hash {} {:#?}", code_hash, err);
-                }
+                if let Err(_err) = &res {}
```
same here

winterjihwan replied:
same as prev reply

djm07073 reviewed (changes requested):
do `rustup run nightly cargo fmt` and `cargo clippy` please

winterjihwan committed:
examples: worker example

winterjihwan commented:
> do `rustup run nightly cargo fmt` and `cargo clippy` please

my bad, I'll not forget it

winterjihwan committed:
refactor: fix unused code, force using arc

winterjihwan committed:
examples: add working worker example

winterjihwan committed:
examples: few bugs and fixes

djm07073 committed:
fix: refactor

- make enum to improve readablity
- delete unused crates import
- clean up test code
- use rwlock more precise way

winterjihwan commented on crates/revmc-worker/src/external.rs:
```diff
@@ -43,28 +49,24 @@ impl EXTCompileWorker {
         {
-            let mut write_lock = match self.cache.try_write() {
-                Ok(g) => g,
+            let cache = match self.cache.try_write() {
```
It will panic if unwraps None, why not just return immediately?

winterjihwan commented on crates/revmc-worker/src/external.rs:
```diff
@@ -16,24 +16,30 @@ use revmc::EvmCompilerFn;
 pub(crate) static SLED_DB: OnceCell<Arc<RwLock<SledDB<B256>>>> = OnceCell::new();
 
+#[derive(PartialEq, Debug)]
+pub enum FetchResult {
```
Personal opinion: naming should be more explicit
suggesion: ExternFnCacheFetchResult

winterjihwan commented on crates/revmc-worker/src/handler.rs:
```diff
@@ -20,26 +18,37 @@ pub fn register_handler<DB: Database + 'static>(
                     context.external.work(spec_id, code_hash, bytecode.original_bytes())
                 {
-                };
+                    tracing::error!("Worker failed: with bytecode hash {}: {:#?}", code_hash, err);
```
Didn't we agree that we will not use tracer?

winterjihwan committed:
chore: fixes, and refactor

winterjihwan committed:
chore: remove unnecessay immports
